rfcs/0042-config-option.md

```diff
@@ -0,0 +1,301 @@
+- Settings that are necessary for the module to work and are different for every user, so they can't have a default. Examples are `services.dd-agent.api_key`, `services.davmail.url` or `services.hydra.hydraURL`. Having a separate option for these settings can give a much better error message when you don't set them (instead of failing at runtime or having to encode the requirement in an assertion) and better discoverability.
+- Password settings: Some program's configuration files require passwords in them directly. Since we try to avoid having passwords in the Nix store, it is advisable to provide a `passwordFile` option as well, which would replace a placeholder password in the configuration file at runtime.
+
+This `config` approach described here is very flexible for these kind of additions, here's an example where we add a `domain` setting to the above service as a separate NixOS option:
+```nix
+{ config, lib, ... }: with lib; {
+
+  options.services.foo = {
+    # ...
+    
+    domain = mkOption {
+      type = types.str;
+      description = "Domain this service operates on.";
+    };
+  };
+  
+  config = mkIf cfg.enable {
+    services.foo.config = {
+      # ...
+      domain = mkDefault cfg.domain;
+    };
+  };
+}
+```
+
+Even though we have two ways of specifying this configuration setting now, the user is free to choose either way.
+
+
+## Configuration checking
+
+One general downside of this approach is that the module system is not able to check the types and values of the configuration file, which would be fast, simple and give good error messages by default. While it would be possible to use `types.addCheck` for the type of the `config` option, this sounds more painful than it's worth and would lead to bad error messages, so we'll ignore this here. Here are some alternatives.
+
+### Configuration checking tools
+
+A number of programs have tools for checking their configuration that don't need to start the program itself. We can use this to verify the configuration at **build time** by running the tool for a derivation build. While this is not as fast as if we had the module system do these checks, which would be at evaluation time already, it is faster than the program failing at runtime due to misconfiguration. These tools however are also more powerful than the module system and can integrate tightly with the program itself, allowing for more thorough checks. In addition, it reduces the amount of RAM needed for evaluation. If a configuration checking tool is available, optimally by the program itself, it should be used if possible, as it can greatly improve user experience. The following illustrates an example of how this might look like
```

Comment:
I'm having two main concerns about the use of checking tools:

1. Inconsistent error messages. An error in the config of module A will be reported in a format completely different from an error in module B. I'm no UX person but this strikes me as giving a poor impression.

2. The difficulty of identifying the location of an error. The checking tool will likely report the line of the error within the generated file, which is next to useless for a user. If I get an configuration error I want it to tell me where among my Nix files I should go to fix it…

Reply:
1. Yeah but I don't think we can do anything against that, having these checks is better than having no checks. Encoding everything as NixOS options does have the advantage that you get uniform type errors at least, but they can't give as good contextual error messages as checking tools, and they of course bring an incredible maintenance burden.

2. Ah that would be neat, it should be possible to do this by using `options.services.foo.config.definitions` to get files where the option is defined. We can add this to the error message in case of check failure. I'll add that.

Reply:
Added a Todo for this in d1a8974c2ffee9453c0b3e2da0bebb75b1e0bcb2

Tbh the scope of this RFC is a bit too wide already though, might have to strip it down a bit.